src/elfxx.c

```diff
@@ -445,6 +445,91 @@ elf_w (find_section) (struct elf_image *ei, const char* secname)
   return 0;
 }
 
+
+static char *
+elf_w (add_hex_byte) (char *str, uint8_t byte)
+{
+  const char hex[] = "0123456789abcdef";
+
+  *str++ = hex[byte >> 4];
+  *str++ = hex[byte & 0xf];
+  *str = 0;
+
+  return str;
+}
+
+
+static int
+elf_w (find_build_id_path) (const struct elf_image *ei, char *path, unsigned path_len)
+{
+  const Elf_W (Ehdr) *ehdr = ei->image;
+  const Elf_W (Shdr) *shdr;
+  unsigned i;
+
+  if (!elf_w (valid_object) (ei))
+    return -1;
+
+  shdr = elf_w (section_table) (ei);
+  if (!shdr)
+    return -1;
+
+  for (i = 0; i < ehdr->e_shnum; ++i, shdr = (const Elf_W (Shdr) *) (((const uint8_t *) shdr) + ehdr->e_shentsize))
```

Comment:
Section headers could be completely stripped out (think core dumps or embedded images). You're better off looking for a PT_NOTE segment through the phdrs since that's guaranteed to be present.